library/src/scala/annotation/MainAnnotation.scala

```diff
@@ -0,0 +1,122 @@
+package scala.annotation
+
+/** MainAnnotation provides the functionality for a compiler-generated main class.
+ *  It links a compiler-generated main method (call it compiler-main) to a user
+ *  written main method (user-main).
+ *  The protocol of calls from compiler-main is as follows:
+ *
+ *    - create a `command` with the command line arguments,
+ *    - for each parameter of user-main, a call to `command.argGetter`,
+ *      or `command.varargGetter` if is a final varargs parameter,
+ *    - a call to `command.run` with the closure of user-main applied to all arguments.
+ *
+ *  Example:
+ *  ```scala
+ *  /** Sum all the numbers
+ *   *
+ *   *  @param first Fist number to sum
+ *   *  @param rest The rest of the numbers to sum
+ *   */
+ *  @myMain def sum(first: Int, rest: Int*): Int = first + rest.sum
+ *  ```
+ *  generates
+ *  ```scala
+ *  object foo {
+ *    def main(args: Array[String]): Unit = {
+ *      val cmd = new myMain().command(
+ *        info = new CommandInfo(
+ *          name = "foo.main",
+ *          documentation = "Sum all the numbers",
+ *          parameters = Seq(
+ *            new ParameterInfo("first", "scala.Int", hasDefault=false, isVarargs=false, "Fist number to sum"),
+ *            new ParameterInfo("rest", "scala.Int" , hasDefault=false, isVarargs=true, "The rest of the numbers to sum")
+ *          )
+ *        )
+ *        args = args
+ *      )
+ *      val args0 = cmd.argGetter[Int](0, None) // using cmd.Parser[Int]
+ *      val args1 = cmd.varargGetter[Int] // using cmd.Parser[Int]
+ *      cmd.run(() => sum(args0(), args1()*))
+ *    }
+ *  }
+ *  ```
+ */
+@experimental
+trait MainAnnotation extends StaticAnnotation:
+  import MainAnnotation.*
+
+  /** The class used for argument string parsing and arguments into a `T` */
+  type Parser[T]
+
+  /** The required result type of the main method.
+   *
+   *  If this type is Any or Unit, any type will be accepted.
+   */
+  type Result
+
+  /** A new command with arguments from `args`
+   *
+   *  @param info The information about the command (name, documentation and info about parameters)
+   *  @param args The command line arguments
+   */
+  def command(info: CommandInfo, args: Array[String]): Command[Parser, Result]
```

Comment:
I'm considering removing this method from the API but still require `MainAnnotation` implementations to define it.

This would provide extra flexibility which would make it possible to define it as a macro. The following example is currently impossible but could work with this change.
```scala
class myMacroAnnot extends  MainAnnotation:
  type Parser[T] = ...
  type Result = ...
  inline def command(inline info: CommandInfo, args: Array[String]): Command[Parser, Result] = ...
```

Reply:
The same could apply to `argGetter` and `varargGetter`.  There we would only need the type parameter and the index parameter and the default option parameter. The implicit parser is not really needed as we could use `summonInline` or the macro equivalent to conditionally try to summon parsers.

```scala
inline def argGetter[T](inline idx: Int, inline defaultArgument: Option[() => T]): () => T = {
  ...
  summonInline[Parser[T]]
  ...
}
```

Reply:
I could make this change in a second PR and discuss the tradeoff there.